sonaranalyzer-dotnet/src/SonarAnalyzer.CSharp/Rules/ImplementISerializableCorrectly.cs

```diff
@@ -0,0 +1,246 @@
+
+                implementationErrors.AddRange(CheckSerializableAttribute(classDeclaration, typeSymbol));
+                implementationErrors.AddRange(CheckConstructor(classDeclaration, typeSymbol));
+                implementationErrors.AddRange(CheckGetObjectDataAccessibility(typeSymbol, getObjectData));
+                implementationErrors.AddRange(CheckGetObjectData(typeSymbol, getObjectData, c.SemanticModel, classDeclaration));
+
+                if (implementationErrors.Count > 0)
+                {
+                    c.ReportDiagnostic(Diagnostic.Create(rule, classDeclaration.Identifier.GetLocation(),
+                        additionalLocations: implementationErrors.ToAdditionalLocations(),
+                        properties: implementationErrors.ToProperties()));
+                }
+            },
+            SyntaxKind.ClassDeclaration);
+        }
+
+        private static IEnumerable<SecondaryLocation> CheckSerializableAttribute(ClassDeclarationSyntax classDeclaration, INamedTypeSymbol typeSymbol)
+        {
+            if (!HasSerializableAttribute(typeSymbol) &&
+                !typeSymbol.IsAbstract)
+            {
+                yield return new SecondaryLocation(classDeclaration.Keyword.GetLocation(),
+                    $"Add 'System.SerializableAttribute' attribute on '{typeSymbol.Name}' because it implements 'ISerializable'.");
+            }
+        }
+
+        private static IEnumerable<TSyntax> GetDeclarations<TSyntax>(ISymbol symbol)
+        {
+            if (symbol == null)
+            {
+                return Enumerable.Empty<TSyntax>();
+            }
+
+            return symbol.DeclaringSyntaxReferences.Select(r => r.GetSyntax()).Cast<TSyntax>();
+        }
+
+        private static IEnumerable<SecondaryLocation> CheckGetObjectDataAccessibility(INamedTypeSymbol typeSymbol, IMethodSymbol getObjectData)
+        {
+            if (!typeSymbol.IsSealed &&
+                getObjectData != null &&
+                !getObjectData.IsOverride &&
+                !(getObjectData.IsVirtual && getObjectData.DeclaredAccessibility == Accessibility.Public))
+            {
+                foreach (var declaration in GetDeclarations<MethodDeclarationSyntax>(getObjectData))
+                {
+                    yield return new SecondaryLocation(declaration.Identifier.GetLocation(),
+                        $"Make this method 'public' and 'virtual', or seal '{typeSymbol.Name}'.");
+                }
+            }
+        }
+
+        private static IEnumerable<SecondaryLocation> CheckGetObjectData(INamedTypeSymbol typeSymbol, IMethodSymbol getObjectData, SemanticModel semanticModel, ClassDeclarationSyntax classDeclaration)
+        {
+            if (!ImplementsISerializable(typeSymbol.BaseType))
+            {
+                yield break;
+            }
+
+            if (getObjectData == null)
+            {
+                var serializableFields = SerializableFieldNames(typeSymbol);

```

Comment:
You are going through the enumerable multiple times so I would suggest calling `.ToList()`